external-crates/move/crates/move-vm-runtime/src/jit/execution/translate.rs

```diff
@@ -282,13 +301,203 @@ pub fn package(
     })
 }
 
+fn modules(
+    package_context: &mut PackageContext<'_>,
+    pkg_module_ids: &BTreeSet<ModuleId>,
+    package_modules: &[input::Module],
+) -> PartialVMResult<()> {
+    use std::collections::BTreeMap;
+
+    macro_rules! make_invariant_violation {
+        ($msg:expr) => {
+            PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR).with_message($msg)
+        };
+    }
+
+    #[derive(Copy, Clone, Eq, PartialEq)]
+    enum State {
+        NotVisited,
+        Visiting,
+        Visited,
+    }
+
+    let input_modules: BTreeMap<ModuleId, &input::Module> = package_modules
+        .iter()
+        .map(|m| (m.compiled_module.self_id(), m))
+        .collect();
+
+    let mut state: BTreeMap<ModuleId, State> = BTreeMap::new();
+
+    for root_id in input_modules.keys() {
+        let root_key = package_context.interner.intern_ident_str(root_id.name())?;
+
+        // Skip if we already fully processed this module.
+        if matches!(state.get(root_id), Some(State::Visited)) {
+            debug_assert!(package_context.loaded_modules.contains_key(&root_key));
+            continue;
+        }
+
+        let mut stack: Vec<ModuleId> = Vec::new();
+        stack.push(root_id.clone());
+
+        while let Some(cur_id) = stack.pop() {
+            let cur_state = *state.get(&cur_id).unwrap_or(&State::NotVisited);
+
+            match cur_state {
+                State::Visited => continue,
+                State::Visiting => {
+                    // all deps done, now load if needed
+                    let key = package_context.interner.intern_ident_str(cur_id.name())?;
+                    if !package_context.loaded_modules.contains_key(&key) {
+                        let input_module = input_modules.get(&cur_id).ok_or_else(|| {
+                            make_invariant_violation!(format!(
+                                "Module {} not found in initial modules",
+                                cur_id
+                            ))
+                        })?;
```

Comment:
Thoughts on adding a check here that all immediate deps are loaded at this point? On the fence on if it makes sense or if that's just too much paranoia....